### PR TITLE
Concept for new rev1 node js client

### PR DESCRIPTION
**Pull request by matthewfl**
Last night I upgraded the [jsonapi-client](https://github.com/matthewfl/jsonapi-node) that I started a few months ago.  This new version uses getters/setters along with promises to make it feel more like the other clients which are able to make synchronous network calls.

Interacting with this version of the client looks like:

``` js
balanced.marketplace.customers.create().add_card(
        balanced.marketplace.cards.create({
            'number': '4111111111111111',
            'expiration_year': '2016',
            'expiration_month': '12'
        })
    ).cards.get(0).debit(500)
     .then(function (debit_obj) { /* everything worked ok */ });
```

Currently everything in the [test](https://github.com/matthewfl/balanced-node/blob/rev1/test/test.js) file works, and the parts that are specific to the resources themselves so far is minimal ([this](https://github.com/matthewfl/balanced-node/blob/rev1/lib/balanced.js))

### CODE REVIEW

**Comment by mjallday** [comment]
@matthewfl may be worth looking at #20

**Comment by matthewfl** [comment]
@mjallday I already had a lot of this, so I just cleaned it a little last night.  I saw that issue, but I didn't see any example clients to figure out what the interaction was like.  

I mostly am trying to clean up the user experience with this library, and make it more like the other clients which basically make blocking http call all over the place and immediately return the object, which is very clean, but does not work with the non blocking nature of node without the abstraction of promises.

**Comment by mjallday** [comment]
@matthewfl i've seen a lot of libraries handle that with `actionAsync(success, failure)` as well as the more traditional `action()` method.

**Comment by matthewfl** [comment]
I haven't experimented with action methods yet with a hypermedia api in node.js.  What I do know is the way it was written before with callbacks (more traditional node fashion) does not appear to work well with hypermedia as each time a request is made it is basically another indent.
eg: https://github.com/balanced/rentmybikes-node/blob/master/card_processing.js#L41-L67
could now become

``` js
var from = balanced.get(from_customer_href);
var to = balanced.get(to_customer_href);
var transactions = to.bank_accounts.get(0).then(function (ba) {
   if(ba) { // we have a bank account
      return from.cards.get(0).debit(amount).thenResolve(ba).credit(amount)
   }
})
```

tldr: the current client doesn't really work well with hypermedia and there needs to be some new way to interact with hypermedia in a non blocking environment.  Whether or not this is the best way to approach this, idk but I am open to other ideas provided examples how how the interface would work.

**Comment by pksunkara** [comment]
@matthewfl An example client library generated using alpaca is at https://github.com/alpaca-api/buffer-node

**Comment by matthewfl** [comment]
@pksunkara ok, thanks, but I am still trying to get a good understanding about what the api would look like when interacting with our api.  Since we are a hypermedia api, it is not uncommon for there to be a number of http get requests before they start preforming an action.  In the old node client, this became really annoying as it was callback based.  

https://github.com/balanced/rentmybikes-node/blob/master/card_processing.js#L41-L67

**Comment by mjallday** [comment]
If this is not a good candidate for trying out alpaca we could look at the [balanced php client](http://github.com/balanced/balanced-php). @pksunkara how's the support for php?

**Comment by pksunkara** [comment]
Same as node. An example is here, https://github.com/alpaca-api/buffer-php

It's installable by composer and PSR-0 compilant. Fully tested the generated library with PHP [ 5.4 / 5.5 ]

**Comment by mjallday** [comment]
@matthewfl how's progress on this?

**Comment by matthewfl** [comment]
@mjallday it works, there might be a few minor bugs that still need to get worked out, but otherwise it should just be getting the scenarios rewritten (as the interface has greatly changed) and more tests

**Comment by mjallday** [comment]
@matthewfl can you spec out the clean up tasks? @tarunc is available to work on this.

**Comment by matthewfl** [comment]
- [ ] Write more test, we shouldn't need the same number of tests as before since this system is more DRY eg: we don't need to explicitly test create method on every resource https://github.com/balanced/balanced-node/blob/master/test/test.js
- [x] Write all the scenarios https://github.com/balanced/balanced-docs/tree/multi-rev/clients/rev1/curl
- [ ] Add to the docs
- [ ] Update the readme to explain the basics such as the system is using promises etc
- [x] publish latest version of jsonapi-node to npm (@matthewfl https://github.com/matthewfl/jsonapi-node )
- [ ] publish this to npm

ps: I guess we should merge this pr so that other can start working on it, as it is currently on my fork

**Comment by mjallday** [comment]
@tarunc - can you take a look at this?

**Comment by matthewfl** [comment]
Going to merge this pr so that the docs can include this as a submodule for generating docs
